### PR TITLE
Email Drafts

### PR DESCRIPTION
**Pull request by dongilbert**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🔴 
| New feature/enhancement? (use the a.x branch)      | 🟢
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🟢 
| Related user documentation PR URL      | mautic/user-documentation#345
| Related developer documentation PR URL | N/A
| Issue(s) addressed                     | N/A

## Description
"Drafts" feature for email from Acquia. User documentation can be found at the link above.

User must enable this feature in the config:
Screenshot 2024-11-19 at 11 50 54

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Ensure email drafts are enabled by modifying the configuration as described above.
3. Create an email with some content, and save it.
4. Edit the email again, and make a change to the content
5. Instead of saving, use the "Save as Draft" button to save the draft.
6. Check both the Preview URL and the Draft Preview URL to see the differences between the emails.
7. Edit the email again, and apply the draft. See that the changes are applied and the Draft Preview URL is no longer present.
8. Edit the email again and create another draft.
9. Discard the draft this time, and see that the changes do not get applied.

### CODE REVIEW

**Review by escopecz — CHANGES_REQUESTED** [review]
@dongilbert please check the failing CI checks

**Comment by dongilbert** [comment]
@escopecz This is failing PHPStan with
```Run if [[ "PHPSTAN" == "PHPSTAN" ]]; then
> [ ! -f var/cache/test/AppKernelTestDebugContainer.xml ] && (echo 'Building test cache ...'; APP_ENV=test APP_DEBUG=1 bin/console > /dev/null 2>&1);  php -d memory_limit=4G bin/phpstan analyse --ansi '--no-progress'
Building test cache ...
Note: Using configuration file /home/runner/work/mautic/mautic/phpstan.neon.
Error: Trying to mock an undefined method findOneBy() on class MauticPlugin\GrapesJsBuilderBundle\Entity\GrapesJsBuilderRepository.
Error: Trying to mock an undefined method findOneBy() on class MauticPlugin\GrapesJsBuilderBundle\Entity\GrapesJsBuilderRepository.
Error: Trying to mock an undefined method findOneBy() on class MauticPlugin\GrapesJsBuilderBundle\Entity\GrapesJsBuilderRepository.
 ------ ---------------------------------------------------------------------------------- 
  Line   plugins/GrapesJsBuilderBundle/Tests/Unit/EventSubscriber/EmailSubscriberTest.php  
 ------ ---------------------------------------------------------------------------------- 
  [71](https://github.com/mautic/mautic/actions/runs/10270839061/job/28419485488?pr=14028#step:10:72)     Trying to mock an undefined method findOneBy() on class                           
         MauticPlugin\GrapesJsBuilderBundle\Entity\GrapesJsBuilderRepository.              
  96     Trying to mock an undefined method findOneBy() on class                           
         MauticPlugin\GrapesJsBuilderBundle\Entity\GrapesJsBuilderRepository.              
  125    Trying to mock an undefined method findOneBy() on class                           
         MauticPlugin\GrapesJsBuilderBundle\Entity\GrapesJsBuilderRepository.              
 ------ ----------------------------------------------------------------------------------```

But I've verified that the method exists in the class hierarchy - any ideas why it may be failing?

File: plugins/GrapesJsBuilderBundle/Tests/Unit/EventSubscriber/EmailSubscriberTest.php
Comment (escopecz) [inline plugins/GrapesJsBuilderBundle/Tests/Unit/EventSubscriber/EmailSubscriberTest.php]: It's weird. One thing I can think of is that the type should be `EmailModel&MockObject` instead but PHP 8.0 doesn't support that yet. Can you try if
```suggestion
    /** @var MockObject&GrapesJsBuilderRepository */
    private MockObject $grapesJsBuilderRepo;
```
would satisfy PHPSTAN?

File: plugins/GrapesJsBuilderBundle/EventSubscriber/EmailSubscriber.php
Comment (dadarya0) [inline plugins/GrapesJsBuilderBundle/EventSubscriber/EmailSubscriber.php:105]: When applying the draft, the value of `DraftCustomMjml` should be transferred to `CustomMjml`, and then `DraftCustomMjml` should be set to null. However, I couldn't find any code that updates `CustomMjml`.

File: plugins/GrapesJsBuilderBundle/EventSubscriber/EmailSubscriber.php
Comment (dadarya0) [inline plugins/GrapesJsBuilderBundle/EventSubscriber/EmailSubscriber.php:100]: There is no need to update `CustomHtml` as we are saving only draft and not changing anything in original email content.

**Review by dadarya0 — CHANGES_REQUESTED** [review]

File: plugins/GrapesJsBuilderBundle/EventSubscriber/EmailSubscriber.php
Comment (dongilbert) [reply plugins/GrapesJsBuilderBundle/EventSubscriber/EmailSubscriber.php:100]: When saving as draft the customHtml of the email entity is updated, this line resets it back to what it was.

File: plugins/GrapesJsBuilderBundle/EventSubscriber/EmailSubscriber.php
Comment (dongilbert) [reply plugins/GrapesJsBuilderBundle/EventSubscriber/EmailSubscriber.php:105]: The customMjml is already up to date when applying the draft. I originally had it as you stated but it is not needed. Please test to verify.

**Comment by dadarya0** [comment]
@dongilbert on save draft getting 500 :-

`500 Internal Server Error - Cannot assign null to property MauticPlugin\GrapesJsBuilderBundle\EventSubscriber\EmailSubscriber::$existingHtml of type string 
`

image

**Comment by dadarya0** [comment]
@dongilbert  Please update PR description and step to test.

I have tested but it is not working as expected.

1. Create an email with Blank theme and save it
2. Edit same email and select `Brienz` template and Click on `Save as Draft` button

**Expected Behaviour :-**
- Email preview should be blank theme
- Draft preview should be `Brienz` theme
- Send email should use blank theme content html

**Actual Behaviour :-**
- Email preview have  Brienz theme
- Draft preview have no content to preview
- Send email is using Brienz theme content html which was save as draft only

**Comment by dongilbert** [comment]
@dadarya0 The customHtml not populating after switching the template _without_ opening the builder is due to an issue with the GrapesJS plugin; the plugin overwrites Mautic's core `setThemeHtml` method, and the endpoint it uses to fetch the MJML for the builder does not return template HTML. Checking to see what I can do to address that.

**Comment by dongilbert** [comment]
@dadarya0 I discovered in the GrapesJS builder service commented out code which handled setting the customHtml from the provided mjml when changing themes. I uncommented and modified it, as well as added code required to support it. This change simulates the same process as opening and closing the builder.

Testing the changes I can confirm actual behavior of creating an email, saving, changing theme to brienz, saving as draft, then sending an example, and sending actual - 
- email preview remains Blank theme
- email draft preview is Brienz
- both example and normal send send Blank theme content

**Comment by dadarya0** [comment]
Thanks @dongilbert , now it is working as expected.

**Review by dadarya0 — DISMISSED** [review]

**Comment by escopecz** [comment]
@dongilbert can you please also update the PR description and prepare docs for this feature in https://github.com/mautic/user-documentation/blob/5.x/docs/channels/emails.rst. I guess https://docs.acquia.com/campaign-studio/channels/emails/email-basics#section-saving-an-email-draft can be used as an inspiration.

**Review by kuzmany — DISMISSED** [review]
From the code I understand the feature.
I fixed one migration issue https://github.com/mautic/mautic/pull/14028/commits/f2a7b295234b982aa8f7bfd585ae4655fa0f945a#diff-0c3bf33b0ca2634d0025d76ddec2dc9a526f94a194d437a9dd467910a7cf2363

I've enabled the feature to insert to local.php     'email_draft_enabled' => 1
THen I was able to make draft for the already existed emails , update drafts, or make it alive.

Good to go 👍

**Comment by dongilbert** [comment]
@escopecz @kuzmany Please review the content changes I made in the PR body as well as the linked documentation.

**Review by escopecz — DISMISSED** [review]
I made some code cleanup and added the option to enable this feature into the UI configuration.

I tested this feature and it worked as described 👍

**Review by escopecz — APPROVED** [review]

**Comment by markusVJH** [comment]
Tested around, worked nicely on my side as well. Cool addition!

File: plugins/GrapesJsBuilderBundle/Assets/library/js/builder.js
Comment (escopecz) [inline plugins/GrapesJsBuilderBundle/Assets/library/js/builder.js:89]: @dongilbert do you know why this was uncommented and changed? It's now blocking the release of M5.2 as it prevents create new emails with error:
```
builder.js?v5c3b11ed:76 Uncaught TypeError: h.getAssetsConfig is not a function
    at Object.success (builder.js?v5c3b11ed:76:11169)
```

Comment (escopecz) [reply plugins/GrapesJsBuilderBundle/Assets/library/js/builder.js:89]: I think it started failing after I merged this automated asset rebuild PR: https://github.com/mautic/mautic/pull/14278

Comment (RCheesley) [reply plugins/GrapesJsBuilderBundle/Assets/library/js/builder.js:89]: https://github.com/mautic/mautic/issues/14299 here's the bug report

Comment (patrykgruszka) [reply plugins/GrapesJsBuilderBundle/Assets/library/js/builder.js:89]: Not sure if this uncommented code is needed, but the builder service should be initialized like this:
```
const assetService = new AssetService();
const builder = new BuilderService(assetService);
```
Do not revive `getAssetsConfig`

**Pull request merged**
